### PR TITLE
VideoPlayer

### PR DESCRIPTION
**Pull request by FernetMenta**
It is impossible to separate distinct work packages from my VideoPlayer branch. Here come almost all changes
- rename dvdplayer to VideoPlayer
- refactoring for hw decoders and renderers
- VideoPlayer owns video renderer
- RenderManager is not global anymore
- audio sync was moved from player into AE
- rework a/v sync
- refactor of switching refresh rates
- refactor of audio passthrough: iex packing is optional and done one if sink requests it
- dithering for GL rendering, credits to @laurimyllari 
- multi video streams by @ace20022 
- dropped libstagefright
- add EGL for X11, zero copy rendering for vaapi

### CODE REVIEW

**Comment by FernetMenta** [comment]
@afedchin VS filters (not sure about project) had merge conflicts I was not able to resolve

**Comment by koying** [comment]
Obviously quite difficult to review, so I'll trust you ;)
How do you want to proceed? Merge and go from there?

**Comment by FernetMenta** [comment]
good question. it depends on the available time of the platform devs. imo it should at least compile for all targets before merge. what do you suggest?

**Comment by hudokkow** [comment]
As it stands, this PR almost crashes the browser when you hit "Files changed" above and then becomes almost unusable...

I say if it builds on all targets and runtime is OK(ish), merge. Last time I tested your branch was over a week ago and it worked. I trust this will too.
Anyway, I'm going to pull this PR and start testing on Ubuntu tonight.

**Comment by ace20022** [comment]
@FernetMenta Could you please or replace my commit with https://github.com/ace20022/xbmc/commit/56ec8165232ef38482b1e87235fe895adb682cbb

There's one wrong function call in it ;)

**Comment by FernetMenta** [comment]
@ace20022 will do

**Comment by FernetMenta** [comment]
@popcornmix could you please verify omxplayer. due to reorder and squash of commits I see a small diff to my current master

**Comment by FernetMenta** [comment]
jenkins build this please

**Comment by afedchin** [comment]
@FernetMenta vs filters looks good. let's wait what jenkins will say

**Comment by koying** [comment]
@FernetMenta Agreed that it should pass jenkins. 
I find it easier to "functionally" fix on master thereafter

**Comment by fritsch** [comment]
Expect PRs for the Windows Sinks - I did those blindly ... and DirectSound will fail (damn braket missing).

Wohooo was a long time.

**Comment by MilhouseVH** [comment]
This is a debug log from latest kodi master (OpenELEC, Nvidia_Legacy) with the latest version of this PR: http://sprunge.us/WbEU

Video playback hangs at startup for almost 2 minutes, before eventually playing, but only at about 2 frames per second (ie. very jerky). This behaviour is the same with or without passthrough audio.

An identical build using your [latest master](https://github.com/xbmc/xbmc/compare/xbmc:fde0686b7845e216c74d0305b803569c6e2288f5...FernetMenta:a99657f89ac1bc3dc24514483d279a5ad905f379) in place of #8486 is working normally, so maybe some commits are missing from #8486?

**Comment by hudokkow** [comment]
I've been playing with this PR for the last 3h and everything is OK. Tried a few movies, TV shows, Youtube videos, etc. It played everything I tried. BTW, fantastic start playing times. Great stuff!
Tested with legacy hardware: C2D T4400@2.20GHz, RV710/M92 [Mobility Radeon HD 4530/4570/545v], 4GB RAM, etc...

**Comment by FernetMenta** [comment]
@MilhouseVH 
my master is rebased on videoplayer branch. the remaining commits are floating on top. most of them are for cmake. I can only think of those that can make a difference:
https://github.com/FernetMenta/xbmc/commit/4132b68c1965a2f9e4c9458db4eba0d836f79d74
https://github.com/FernetMenta/xbmc/commit/c21fe9a63e8c17bc6727591091024177ea5f6a2a

**Comment by fritsch** [comment]
I sent a PR for the Wasapi build failure ... we need someone with windows that can actually test the resulting Sink ... as some structs have changed. I am not totally sure I did not miss something or set wrong values from time to time ...

**Comment by MilhouseVH** [comment]
>for cmake. I can only think of those that can make a difference:
> FernetMenta@4132b68
> FernetMenta@c21fe9a

Yep, with both of those playback is working - let me narrow down which is required.

**Comment by MilhouseVH** [comment]
I need the PTHREADS commit: FernetMenta@4132b68 - with just this added, playback is working.

My setup: OpenELEC master, NVidia_Legacy (Revo 3700 with ION2), Onkyo TX-NR828 AVR.

**Comment by FernetMenta** [comment]
@Memphiz @wsnipex the Linux VM is missing libEGL. Could you add it please?

**Comment by fritsch** [comment]
Let's see what happens: http://jenkins.kodi.tv/job/WIN-32/6881/

**Comment by wsnipex** [comment]
I've tried my best to get the needed mesa libs installed on the slave. If it doesn't work, we'll have to set up a new one as we're still on ubuntu 12.04...

**Comment by wsnipex** [comment]
jenkins build this please

**Comment by ace20022** [comment]
@FernetMenta please drop my commit on the next occasion

**Comment by FernetMenta** [comment]
@ace20022 done

jenkins build this please

**Comment by FernetMenta** [comment]
> ld: library not found for -lpython2.6

@wsnipex ^^ where does this come from?

**Comment by FernetMenta** [comment]
@wsnipex my guess is that buildsteps do not force make xcode_depends and we toggle between Jarvis branch and master

**Comment by koying** [comment]
@FernetMenta There's another one: https://github.com/koying/xbmc/commit/99dd50a4a07b36f8640d7456b64dd749159cc99a

**Comment by FernetMenta** [comment]
@koying thanks
@wsnipex disregard, merge issue here

**Comment by FernetMenta** [comment]
jenkins build this please

**Comment by koying** [comment]
@FernetMenta Please pick https://github.com/koying/xbmc/commit/01be58f64f44cd3d21c2fd08833e8201b0c53f93 for mediacodec surface

**Comment by hudokkow** [comment]
Android X86 failed with
> In file included from platform/Implementation.cpp:22:0:
/home/jenkins/workspace/Android-X86/xbmc/threads/platform/pthreads/Implementation.cpp: In function 'bool XbmcThreads::pthreads::setRecursiveAttr()':
/home/jenkins/workspace/Android-X86/xbmc/threads/platform/pthreads/Implementation.cpp:42:54: error: 'PTHREAD_PRIO_INHERIT' was not declared in this scope
         pthread_mutexattr_setprotocol(&recursiveAttr,PTHREAD_PRIO_INHERIT);
                                                      ^
/home/jenkins/workspace/Android-X86/xbmc/threads/platform/pthreads/Implementation.cpp:42:74: error: 'pthread_mutexattr_setprotocol' was not declared in this scope
         pthread_mutexattr_setprotocol(&recursiveAttr,PTHREAD_PRIO_INHERIT);
                                                                          ^

Win32 didn't even run. new test build: http://jenkins.kodi.tv/job/WIN-32/6891/

**Comment by koying** [comment]
pthread_mutexattr_setprotocol not in NDK altogether, so needs to exclude android as a whole, not just arm

**Comment by FernetMenta** [comment]
> pthread_mutexattr_setprotocol not in NDK altogether

yes, already done but forgot to add on the last run

**Comment by FernetMenta** [comment]
jenkins build this please

**Comment by MartijnKaijser** [comment]
all passed :)

File: xbmc/cores/AudioEngine/Interfaces/AEStream.h [hunk 1]
```diff
@@ -29,6 +29,32 @@ extern "C" {
...
+public:
+  double delay;
+  double error;
+  double rr;
+  unsigned int errortime;
+  enum AESyncState
+  {
+    SYNC_OFF,
+    SYNC_PLAY,
+    SNYC_ACTIVE
```
Comment (anaconda) [inline xbmc/cores/AudioEngine/Interfaces/AEStream.h]: Typo

**Comment by popcornmix** [comment]
I'm okay to disable it on Pi1. Just checking if an ifdef or a settings option is preferred.
We could leave the setting in place but make it enabled/not-visible by default.

**Comment by FernetMenta** [comment]
> We could leave the setting in place but make it enabled/not-visible by default.

that is like dead code and an abuse of settings. the setting has to go

**Comment by FernetMenta** [comment]
So finally this builds on all targets. I forced pushed after:

- fixed a typo, thanks @anaconda 
- dropped a commit about setting of dca codec, <- @popcornmix 
- fixed a/v sync of passthrough audio, @MilhouseVH 

any objections to get this in after a final build?

**Comment by popcornmix** [comment]
No objections from me.

**Comment by fritsch** [comment]
Also fine by me as well. It was acked by all towers and contains. Will need
further fixes for audio on Windows and work for android though.
Am 06.12.2015 9:22 nachm. schrieb "popcornmix" <notifications@github.com>:

**Comment by FernetMenta** [comment]
jenkins build this please

**Comment by ace20022** [comment]
Just a note and without a log, platform windows: There's a picture freeze and a very high memory consumption during rw/ff 8/16/32x with this test file: http://ftp.nluug.nl/pub/graphics/blender/demo/movies/ToS/tearsofsteel_4k.mov But there's no hang, playback resumes after pressing play.

**Comment by MilhouseVH** [comment]
I have a reproducible crash with this latest PR.

1. Enable passthrough
2. Enable "Dolby Digital (AC3) capable receiver"
3. Play a video with AC3 soundtrack
4. Seek within the video (ie. jump to 99 seconds with `9-9-ENTER`)
5. Kodi will crash a second or two after completing the seek

I can reproduce this on both x86 and RPi2 (both OpenELEC).

Crashlog from RPi2: http://sprunge.us/WiQU (debug-enabled build)
Crashlog from x86 (Nvidia_Legacy, Revo3700 w/ION2): http://sprunge.us/QfJJ (not debug-enabled build)

**Comment by koying** [comment]
Congrats :)

Le lun 7 déc. 2015 07:38, Rainer Hochecker <notifications@github.com> a
écrit :

**Comment by FernetMenta** [comment]
hehe :)
I want to mention this again here: we know that there is still work to do but it is much easier to do this on here on master instead of another branch that would diverge even more.

**Comment by fritsch** [comment]
Yeah - that's what we can do in master now. But let's not forget on those outstanding issue for v16.

**Comment by MilhouseVH** [comment]
Just to say the AC3 crash is fixed in latest build, possibly because of https://github.com/FernetMenta/xbmc/issues/340

**Comment by FernetMenta** [comment]
@popcornmix friendly reminder for an alternative to https://github.com/FernetMenta/xbmc/commit/67eb6156e611d5e249a1962ce60c14d551e8b5f5
I won't forget this :)

**Comment by koying** [comment]
I'm all to drop the setting, but we have to check whether channel layout > 2.0, or it makes no sense imo

**Comment by popcornmix** [comment]
@FernetMenta okay, I'll have a look.

**Pull request merged**